src/MongoDB.Driver.Core/Core/Misc/ExceptionMapper.cs

```diff
@@ -118,17 +118,37 @@ public static Exception Map(ConnectionId connectionId, WriteConcernResult writeC
         /// <returns>The exception, or null if no exception necessary.</returns>
         public static Exception MapNotPrimaryOrNodeIsRecovering(ConnectionId connectionId, BsonDocument response, string errorMessageFieldName)
         {
+            BsonValue codeBsonValue;
+            if (response.TryGetValue("code", out codeBsonValue) && codeBsonValue.IsNumeric)
+            {
+                var code = (ServerErrorCode)codeBsonValue.ToInt32();
+                switch (code)
+                {
+                    case ServerErrorCode.NotMaster:
+                    case ServerErrorCode.NotMasterNoSlaveOk:
+                        return new MongoNotPrimaryException(connectionId, response);
+
+                    case ServerErrorCode.InterruptedAtShutdown:
+                    case ServerErrorCode.InterruptedDueToReplStateChange:
+                    case ServerErrorCode.NotMasterOrSecondary:
+                    case ServerErrorCode.PrimarySteppedDown:
+                    case ServerErrorCode.ShutdownInProgress:
+                        return new MongoNodeIsRecoveringException(connectionId, response);
+                }
+            }
+
```

Comment:
If the new code above fails to map to an exception based on the code we fall through to the string based method below.